datadog_checks_dev/datadog_checks/dev/tooling/commands/test.py

```diff
@@ -132,72 +132,73 @@ def test(
-                        abort('\nFailed!', code=result.code)
-
-                    fix_coverage_report(check, 'coverage.xml')
-                    run_command(['codecov', '-X', 'gcov', '--root', root, '-F', check, '-f', 'coverage.xml'])
+        for env in envs:
+            # This is for ensuring proper spacing between output of multiple checks' tests.
+            # Basically this avoids printing a new line before the first check's tests.
+            output_separator = '\n' if tests_ran else ''
+
+            # For performance reasons we're generating what to test on the fly and therefore
+            # need a way to tell if anything ran since we don't know anything upfront.
+            tests_ran = True
+
+            if coverage:
+                test_env_vars['PYTEST_ADDOPTS'] = pytest_options.format(pytest_coverage_sources(check))
+
+            if verbose:
+                echo_info('pytest options: `{}`'.format(test_env_vars['PYTEST_ADDOPTS']))
+
+            with chdir(os.path.join(root, check), env_vars=test_env_vars):
+                if format_style:
+                    test_type_display = 'the code formatter'
+                elif style:
+                    test_type_display = 'only style checks'
+                elif bench:
+                    test_type_display = 'only benchmarks'
+                elif e2e:
+                    test_type_display = 'only end-to-end tests'
                 else:
+                    test_type_display = 'tests'
+
+                wait_text = '\n{}Running {} for {} : {}'.format(output_separator, test_type_display, check, env)
+                echo_waiting(wait_text)
+                echo_waiting('-' * len(wait_text) + '\n')
+
+                command = (
+                    'tox '
+                    # so users won't get failures for our possibly strict CI requirements
+                    '--skip-missing-interpreters '
+                    # so coverage tracks the real locations instead of .tox virtual envs
+                    '--develop '
+                    # comma-separated list of environments
+                    '-e {}'.format(env)
+                )
+                if env in STYLE_ENVS:
+                    result = run_command(command=command)
+                else:
+                    result = run_command_with_retry(retry=retry, command=command)
+
+                if result.code:
+                    abort('\nFailed!', code=result.code)
+
+                if coverage and file_exists('.coverage'):
                     if not cov_keep:
-                        remove_path('.coverage')
-                        remove_path('coverage.xml')
+                        echo_info('\n---------- Coverage report ----------\n')
+
+                        result = run_command('coverage report --rcfile=../.coveragerc')
+                        if result.code:
+                            abort('\nFailed!', code=result.code)
+
+                    if testing_on_ci:
```

Comment:
Coverage stuff must be after all envs

Reply:
1.
> Can we make this a flag?

You mean running each tox env one by one VS all at once ?
Is there any reason we need that flag, what's the use case ?

2.
>  Coverage stuff must be after all envs

Refactor by moving down the envs loop. But not sure if running tox multiple time before coverage commands will break it. I will make some tests.

Reply:
1. Because for the common use case of everything passing (usually local development) this will be far slower

Reply:
It's indeed somewhat slower:

```
# all at once
ddev test mysql -k test__get_server_pid  18.69s user 5.92s system 92% cpu 26.642 total
ddev test mysql -k test__get_server_pid  18.65s user 6.20s system 92% cpu 26.806 total
ddev test mysql -k test__get_server_pid  18.47s user 6.20s system 92% cpu 26.644 total

# one by one
ddev test mysql -k test__get_server_pid  21.70s user 6.95s system 91% cpu 31.145 total
ddev test mysql -k test__get_server_pid  21.41s user 6.84s system 93% cpu 30.318 total
ddev test mysql -k test__get_server_pid  21.35s user 6.79s system 93% cpu 30.229 total
```

Reply:
Since it's slower. Maybe this PR is not needed, and we can keep retrying for all envs as done today.

Successful builds are far more common than failure. By running tox envs one by one, we are going to slow down all builds.

Reply:
Good catch,
If I understand correctly, this PR adds around ~500ms per tox environment. So for sure, it's a bit slower, but it's not that bad honestly.

That being said, we don't need this if we don't use the `--retry` flag. Maybe we can use that flag? So the impact will only be limited to the CI

Reply:
For a typical check, the cost would be like 3 seconds (~500ms * 3 tox env per check * 2 for py2-3)

In one hand this PR will introduce a penalty of 3 seconds per check on CI (with `--retry`) for all builds.

On the other hand the benefit (time gain) introduce by this PR will only be visible when there is a failure. The rate of failure is much lower than successful build.

This PR + new flag will had complexity. Adding complexity for something that have negative or low benefit (arguments above) doesn't sounds to be a good trade off.

I would favour the current implementation where we retry all tox envs at once :)